plutus-use-cases/src/Plutus.hs

```diff
@@ -140,11 +179,45 @@ data Range a =
 
 -- | Event triggers the Plutus client can register with the wallet.
 data EventTrigger =
-    BlockHeight (Range Int) -- ^ True when the block height is within the range
-    | FundsAtAddress Address (Range Value) -- ^ True when the funds at an address are within the range; TODO: [Address] to add up funds from multiple addresses
+    BlockHeightRange !(Range Int) -- ^ True when the block height is within the range
+    | FundsAtAddress !Address !(Range Value) -- ^ True when the funds at an address are within the range; TODO: [Address] to add up funds from multiple addresses
     | And EventTrigger EventTrigger -- ^ True when both triggers are true
     | Or EventTrigger EventTrigger -- ^ True when at least one trigger is true
     | PAlways -- ^ Always true
     | PNever -- ^ Never true
 
-type Validator = PlutusTx
+-- | Validator scripts expect two scripts and information about the current
+--   txn. In the future this will be written in Plutus (with the help of TH)
```

Comment:
I don't think this is going to work. The Plutus you're going to get is not going to have a Haskell type that lines up with this, it's going to be a blob of code.

Reply:
Why not? The [docs](https://github.com/input-output-hk/plutus-prototype/tree/master/docs/extended-utxo#extension-to-validator-scripts) say

> An extended validator script expects four arguments:

and

> We consider a validator script to have executed successful if it does not terminate in the Plutus Core error state.

So that should its signature, shouldn't it? You can probably remove the type parameter from `PlutusTx`, and the newtype wrapper `Validator`, and that data types might be slightly different, but in principle that should be what a validation script looks like.

Reply:
That's the type *in Plutus*. But the Plutus type system does not magically join up with the Haskell one: a Plutus program of type `Int` and one of type `Int->Int` are both just going to look like `PlcCode` in the coordination script. Now, maybe we could come up with a way to do this in a more typed way, but afaik we're not planning to.

Reply:
Yes but with this code I am trying to emulate the experience of writing validator scripts in Haskell, in order for them to be translated by the plugin. Somewhere there has to be a piece of Haskell code with this signature, or something equivalent.

(Just to be clear, the point of this definition is not for the coordination code to actually evaluate `runValidator`, or to turn PlcCode into Haskell definitions. It's only there to guide the implementation of validation scripts)

Can you maybe give an example of a validator script?

Reply:
You can't reference arbitrary Haskell defined *outside* the splice, because otherwise we could drag in the world. So you might be able to define a local function with this type inside a splice, but you're not going to be able to pull in a top-level binding. So from the point of view of the coordination code all the scripts are just going to be blobs.

(Perhaps this is the confusion: the plugin converts Haskell *expressions* not Haskell *values*. You can't stick a function in a datastructure and then later convert it, that's not how it works.)

I expect a validator script to look more like the example that you commented out in the other file - a function defined *inside* a splice.

Furthermore, the Plutus script isn't going to take something of type `Redeemer`, it's presumably going to take something with the actual type of the redeemer that this script expects. Similarly for the data, which could be anything. My contract might have `Int`s as data, yours might have something else entirely. You're going to want to share the types for those, yes, but on a per-application basis, since they're likely going to be different for each. The generic code that you're writing here doesn't have the information to have properly typed Plutus scripts, tbh!

Reply:
Yeah my plan is to change the code that says `mkValidator . Validator` to `plc` and adjust everything else accordingly (in `Contracts.CrowdFunding.contributionScript`). The reason I commented out the earlier example is that it was not type checked, and I wanted to make sure that the actual script made sense in terms of the data it has access to, because that wasn't entirely clear to me before.

> The generic code that you're writing here doesn't have the information to have properly typed Plutus scripts, tbh!

Yes that is also true, and that's why I wrote `type Redeemer = String` as a stand-in type. It's all made to support finishing the crowdfunding example. But I think it will be best to actually use the plugin here, which is what I'm going to do next.

Reply:
@michaelpj I made some wrong assumptions about how the parts will fit together and about the different stages involved here. Sorry for the confusion. 

I wrote a tentative version of the code that uses the plugin. Is this closer to what you had in mind?

Reply:
I'm actually not sure which one of us is confused - but your new version does look much more like what I was expecting! I'm happy for you to write the normal Haskell version for now so it gets typechecked - I hadn't realised it was supposed to be schematic.

I don't think you should try too hard to use the plugin yet until it's merged. If you do, it's worth noting that you can't use arbitrary Haskell functions from outside the splice unless you can evaluate the whole thing at compile time. So you'll be better off writing it in as "elementary" Haskell as possible.